src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            addFirst(value, head);
+            size++;
+            return;
+        }
+        currentNode = getNodeByIndex(index);
+        Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+        currentNode.prev.next = newNode;
+        currentNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkExistingNode(index);
+        currentNode = getNodeByIndex(index);
+        return currentNode.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkExistingNode(index);
+        currentNode = getNodeByIndex(index);
+        T oldItem = currentNode.item;
+        currentNode.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkExistingNode(index);
+        currentNode = getNodeByIndex(index);
+        unlink(currentNode);
+        size--;
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        currentNode = head;
+        while (currentNode != null) {
+            if (currentNode.item == object || currentNode.item.equals(object)) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
+    private void unlink(Node<T> currentNode) {
+        if (isHead(currentNode)) {
+            removeHead(currentNode);
+            return;
+        }
+        if (isLast(currentNode)) {
+            removeTail(currentNode);
+            return;
+        }
+        currentNode.prev.next = currentNode.next;
+        currentNode.next.prev = currentNode.prev;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void removeTail(Node<T> currentNode) {
+        currentNode.prev.next = null;
+    }
+
+    private void removeHead(Node<T> currentNode) {
+        head = currentNode.next;
+    }
+
+    private void addFirst(T value, Node<T> oldFirst) {
+        Node<T> newNode = new Node<>(null, value, oldFirst);
+        if (oldFirst != null) {
+            oldFirst.prev = newNode;
+        }
+        head = newNode;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> currentNode = head;
+        int indexCurrentNode = 0;
+        while (currentNode != null && indexCurrentNode != index) {
```

Comment:
let's move to the node from the head (if the node is in the first part of the list) or from the tail (if the node is at the second part of the list)